flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/join/stream/state/MultiJoinStateViews.java

```diff
@@ -0,0 +1,440 @@
+
+            this.recordState = ctx.getMapState(recordStateDesc);
+        }
+
+        private RowData createCompositeKey(RowData joinKey, RowData record) {
+            GenericRowData compositeKey = new GenericRowData(2);
+            compositeKey.setField(0, joinKey);
+            compositeKey.setField(1, record);
+            return compositeKey;
+        }
+
+        @Override
+        public void addRecord(RowData joinKey, RowData record) throws Exception {
+            // Normalize RowKind for consistent state representation
+            RowKind originalKind = record.getRowKind();
+            record.setRowKind(RowKind.INSERT);
+            RowData compositeKey = createCompositeKey(joinKey, record);
+
+            Integer currentCount = recordState.get(compositeKey);
+            if (currentCount == null) {
+                currentCount = 0;
+            }
+            recordState.put(compositeKey, currentCount + 1);
+            record.setRowKind(originalKind); // Restore original RowKind after key creation
+        }
+
+        @Override
+        public void retractRecord(RowData joinKey, RowData record) throws Exception {
+            // Normalize RowKind for consistent state representation and lookup
+            RowKind originalKind = record.getRowKind();
+            record.setRowKind(RowKind.INSERT);
+            RowData compositeKey = createCompositeKey(joinKey, record);
+
+            Integer currentCount = recordState.get(compositeKey);
+            if (currentCount != null) {
+                if (currentCount > 1) {
+                    recordState.put(compositeKey, currentCount - 1);
+                } else {
+                    recordState.remove(compositeKey);
+                }
+            }
+            record.setRowKind(originalKind); // Restore original RowKind
+        }
+
+        @Override
+        public Iterable<RowData> getRecords(RowData joinKey) throws Exception {
+            Iterator<Map.Entry<RowData, Integer>> stateIterator = recordState.iterator();
+            if (stateIterator == null) {
+                return Collections.emptyList();
+            }
+
+            //noinspection NullableProblems // TODO Gustavo can we not supress?
+            return new IterableIterator<>() {
+                private RowData currentRecord = null;
+                private int remainingTimes = 0;
+
+                @Override
+                public boolean hasNext() {
+                    if (remainingTimes > 0) {
+                        return true;
+                    }
+                    while (stateIterator.hasNext()) {
+                        Map.Entry<RowData, Integer> currentEntry = stateIterator.next();
+                        RowData compositeKey = currentEntry.getKey();
+                        RowData currentJoinKey = compositeKey.getRow(0, joinKeyType.getArity());
+                        if (joinKey.equals(currentJoinKey)) {
```

Comment:
Here, we assume that `equals` yields the same as equality predicate in JOIN. 
Similarly, we assume that for binary forms of keys (when we use them with RocksDB in map state).
Does this assumption always hold? 


cc: @dawidwys

Reply:
Running the tests with RocksDB should be easy; and I'd say it's mandatory to run the tests with the both state backends (the usual way to do that is to parameterize them).

I can look deeper into it later; I would also be great to hear Dawids opinion

Reply:
We've synced and I'm adding parameterized tests for rocksdb and changing how we check the joinKey equality to make it general 👍

Reply:
Awesome, thanks!